src/main/java/com/snowplowanalytics/redash/RedashClient.java

```diff
@@ -0,0 +1,437 @@
+    private static final String MESSAGE = "message";
+    private static final String NULL = "null";
+    private static final String MESSAGE_URL_NOT_FOUND = "The requested URL was not found on the server.  " +
+                                            "If you entered the URL manually please check your spelling and try again.";
+    private static final String MESSAGE_INTERNAL_SERVER_ERROR = "Internal Server Error";
+    private static final String JSON_CONTENT_TYPE = "application/json; charset=utf-8";
+
+    private static final MediaType JSON = MediaType.parse(JSON_CONTENT_TYPE);
+
+    public static final String DATA_SOURCE_ALREADY_EXISTS = "Data-source with this name already exists.";
+    public static final String USER_GROUP_ALREADY_EXISTS = "User group with this name already exists.";
+    public static final String USER_DOES_NOT_EXIST = "User with such name does not exist.";
+    public static final String DATA_SOURCE_DOES_NOT_EXIST = "Data-source with such name does not exist.";
+
+    private final OkHttpClient client;
+    private final String baseUrl;
+    private final String apiKey;
+    private final Headers headers;
+
+    public RedashClient(String schema, String host, int port, String apiKey) {
+        this.client = new OkHttpClient();
+        this.apiKey = apiKey;
+        this.baseUrl = schema + "://" + host + ":" + port + API_PREFIX;
+        this.headers = new Headers.Builder()
+                .add("Accept", "application/json, text/plain, */*")
+                .add("Content-Type", JSON_CONTENT_TYPE)
+                .build();
+    }
+
+    /**
+     * Creates new data-source instance on the Redash server.
+     *
+     * @param dataSource data transfer object which should contain all necessary information.
+     * @return int id of successfully (if so) created instance. In this case object that was transferred as argument
+     * receives that id.
+     * @throws IOException              if server is unable due to connection error or API key is invalid.
+     * @throws IllegalArgumentException if there is already presents data-source on Redash server
+     *                                  with such name
+     */
+    public int createDataSource(DataSource dataSource) throws IOException, IllegalArgumentException {
+        if (isEntityAlreadyExists(getDataSources(), dataSource.getName())) {
+            throw new IllegalArgumentException(DATA_SOURCE_ALREADY_EXISTS);
+        }
+        String url = baseUrl + DATA_SOURCES_URL_PREFIX + API_KEY_URL_PARAM + apiKey;
+        String response = post(url, new Gson().toJson(dataSource), CheckResponseStatus.NO);
+        int id = getIdFromJson(response);
+        dataSource.setId(id);
+        return id;
+    }
+
+    /**
+     * Updates data-source on Redash server with provided as argument DataSource instance. Please note that password field
+     * doesn't take a part in making decision of necessity to update data-source as the server does not return it (password field).
```

Comment:
In this case how are we ensuring that we update the password when we want to and not updating the password if that part of the upgrade is not required?

Reply:
So if the server sends password as set of "-" symbols, I think there's no way to decide do we need  the update this field or not. In the case if one of fields needs to be updated then the entire entity will be overwritten, with password. If not, it will not. I will fix javadoc with these strings. I can propose to change the method's logic to update entity independently of is it up to date or not.

Reply:
So I am more referring to the need to potentially update the password - how is it signalled that we want to update the password?

One way I guess would be to check if the password field is null in the supplied object and only if it contains a non-empty string to then update the password for this data-source.

There may well be a need to update a password so I want to make sure we can if we need to!

Reply:
I'm sorry but I can't understand:  if the password is null (as user doesn't want to change it), and user wants to change some another feild(s), so what will be the value of "password " field of dto which will be sent onto redash? If null it will fall with BAD REQUEST, if empty string so the empty string becomes new password.
In my opinion user should always provide correct password for updating. Also we can implement merging with non-password fields: if user provided null beside of certain field so it will stay the same.